### ISSUE TITLE
"Foreign keys are both mapped to <Table> but referencing different principal tables" on long entity names

### ISSUE THREAD

**Issue by reinux** [dotnet/efcore#25456]
## Ask a question

I'm getting this exception:

```
Error: System.InvalidOperationException: The foreign keys {'AReallyReallyReallyReallyReallyReallyReallyReallyReallyLongDetailsId'} on 'Location' and {'AReallyReallyReallyReallyReallyReallyReallyReallyReallyLongOwnerEntityId'} on 'AReallyReallyReallyReallyReallyReallyReallyReallyReallyLongDetails' are both mapped to 'AReallyReallyReallyReallyReallyReallyReallyReallyReallyLongDet~.FK_AReallyReallyReallyReallyReallyReallyReallyReallyReallyLong~' but referencing different principal tables ('AReallyReallyReallyReallyReallyReallyReallyReallyReallyLongDet~' and 'Owners').
```

This seems to arise there's a one-to-many relationship, with an owned type on the _one_ side, and when the entity names are long enough to cause that `~` to show up in the error. I think it's checking only the first part of the name for equivalence.

### Include your code
```
  [Owned]
  public class Location
  {
    public float Long { get; set; }
    public float Lat { get; set; }
  }


  public class AReallyReallyReallyReallyReallyReallyReallyReallyReallyLongOwnerEntity
  {
    public int Id { get; set; }
    public IQueryable<AReallyReallyReallyReallyReallyReallyReallyReallyReallyLongDetails> Details { get; set; }
  }

  public class AReallyReallyReallyReallyReallyReallyReallyReallyReallyLongDetails
  {
    public int Id { get; set; }
    public int OwnerId { get; set; }
    public virtual Location Location { get; set; }
  }

  public class DataContext : DbContext
  {
    public virtual DbSet<AReallyReallyReallyReallyReallyReallyReallyReallyReallyLongOwnerEntity> Owners { get; set; }
  }
```

### Include stack traces

```
at Microsoft.EntityFrameworkCore.Metadata.Internal.RelationalForeignKeyExtensions.AreCompatible(IForeignKey foreignKey, IForeignKey duplicateForeignKey, StoreObjectIdentifier& storeObject, Boolean shouldThrow)
at Microsoft.EntityFrameworkCore.Infrastructure.RelationalModelValidator.ValidateCompatible(IForeignKey foreignKey, IForeignKey duplicateForeignKey, String foreignKeyName, StoreObjectIdentifier& storeObject, IDiagnosticsLogger`1 logger)
at Microsoft.EntityFrameworkCore.Infrastructure.RelationalModelValidator.ValidateSharedForeignKeysCompatibility(IReadOnlyList`1 mappedTypes, StoreObjectIdentifier& storeObject, IDiagnosticsLogger`1 logger)
at Microsoft.EntityFrameworkCore.Infrastructure.RelationalModelValidator.ValidateSharedTableCompatibility(IModel model, IDiagnosticsLogger`1 logger)
at Microsoft.EntityFrameworkCore.Infrastructure.RelationalModelValidator.Validate(IModel model, IDiagnosticsLogger`1 logger)
at Npgsql.EntityFrameworkCore.PostgreSQL.Infrastructure.NpgsqlModelValidator.Validate(IModel model, IDiagnosticsLogger`1 logger)
at Microsoft.EntityFrameworkCore.Metadata.Conventions.ValidatingConvention.ProcessModelFinalized(IModel model)
at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.ImmediateConventionScope.OnModelFinalized(IModel model)
at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.OnModelFinalized(IModel model)
at Microsoft.EntityFrameworkCore.Metadata.Internal.Model.FinalizeModel()
at Microsoft.EntityFrameworkCore.ModelBuilder.FinalizeModel()
at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.CreateModel(DbContext context, IConventionSetBuilder conventionSetBuilder, ModelDependencies modelDependencies)
at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.GetModel(DbContext context, IConventionSetBuilder conventionSetBuilder, ModelDependencies modelDependencies)
at Microsoft.EntityFrameworkCore.Internal.DbContextServices.CreateModel()
at Microsoft.EntityFrameworkCore.Internal.DbContextServices.get_Model()
at Microsoft.EntityFrameworkCore.Infrastructure.EntityFrameworkServicesBuilder.<>c.<TryAddCoreServices>b__7_3(IServiceProvider p)
at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitFactory(FactoryCallSite factoryCallSite, RuntimeResolverContext context)
at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
at Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.<>c__DisplayClass1_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
at Microsoft.EntityFrameworkCore.DbContext.get_DbContextDependencies()
at Microsoft.EntityFrameworkCore.DbContext.get_InternalServiceProvider()
at Microsoft.EntityFrameworkCore.DbContext.Microsoft.EntityFrameworkCore.Infrastructure.IInfrastructure<System.IServiceProvider>.get_Instance()
at Microsoft.EntityFrameworkCore.Infrastructure.Internal.InfrastructureExtensions.GetService[TService](IInfrastructure`1 accessor)
at Microsoft.EntityFrameworkCore.Infrastructure.AccessorExtensions.GetService[TService](IInfrastructure`1 accessor)
at Microsoft.EntityFrameworkCore.Infrastructure.DatabaseFacade.get_Dependencies()
at Microsoft.EntityFrameworkCore.Infrastructure.DatabaseFacade.EnsureDeleted()
at FSI_0005.scrub()
at <StartupCode$FSI_0006>.$FSI_0006.main@()
```

### Include provider and version information

EF Core version: 5.0.8
Database provider: Npgsql.EntityFrameworkCore.PostgreSQL 5.0.8
Target framework: .NET 5.0
Operating system: Windows 10 Pro 21H1 (19043.1110)
IDE: Visual Studio 2019

**Comment by reinux** [comment]
Sorry, I wrote this at first thinking it would be a question rather than a bug report. Could it be reclassified?

**Comment by ajcvickers** [comment]
@reinux I am not able to reproduce this--see my code and the model it generates below. Please attach a small, runnable project or post a small, runnable code listing that reproduces what you are seeing so that we can investigate.

```C#
[Owned]
public class Location
{
    public float Long { get; set; }
    public float Lat { get; set; }
}

public class AReallyReallyReallyReallyReallyReallyReallyReallyReallyLongOwnerEntity
{
    public int Id { get; set; }
    public IQueryable<AReallyReallyReallyReallyReallyReallyReallyReallyReallyLongDetails> Details { get; set; }
}

public class AReallyReallyReallyReallyReallyReallyReallyReallyReallyLongDetails
{
    public int Id { get; set; }
    public int OwnerId { get; set; }
    public virtual Location Location { get; set; }
}

public class SomeDbContext : DbContext
{
    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        => optionsBuilder
            .LogTo(Console.WriteLine, LogLevel.Information)
            .UseSqlServer(Your.ConnectionString)
            .EnableSensitiveDataLogging();

    public DbSet<AReallyReallyReallyReallyReallyReallyReallyReallyReallyLongOwnerEntity> Owners { get; set; }
}

public class Program
{
    public static void Main()
    {
        using (var context = new SomeDbContext())
        {
            context.Database.EnsureDeleted();
            context.Database.EnsureCreated();

            var model = context.Model;
        }
    }
}

```

```
Model: 
  EntityType: AReallyReallyReallyReallyReallyReallyReallyReallyReallyLongDetails
    Properties: 
      Id (int) Required PK AfterSave:Throw ValueGenerated.OnAdd
      AReallyReallyReallyReallyReallyReallyReallyReallyReallyLongOwnerEntityId (no field, Nullable<int>) Shadow FK Index
      OwnerId (int) Required
    Navigations: 
      Location (Location) ToDependent Location
    Keys: 
      Id PK
    Foreign keys: 
      AReallyReallyReallyReallyReallyReallyReallyReallyReallyLongDetails {'AReallyReallyReallyReallyReallyReallyReallyReallyReallyLongOwnerEntityId'} -> AReallyReallyReallyReallyReallyReallyReallyReallyReallyLongOwnerEntity {'Id'} ToDependent: Details ClientSetNull
    Indexes: 
      AReallyReallyReallyReallyReallyReallyReallyReallyReallyLongOwnerEntityId 
  EntityType: AReallyReallyReallyReallyReallyReallyReallyReallyReallyLongOwnerEntity
    Properties: 
      Id (int) Required PK AfterSave:Throw ValueGenerated.OnAdd
    Navigations: 
      Details (IQueryable<AReallyReallyReallyReallyReallyReallyReallyReallyReallyLongDetails>) Collection ToDependent AReallyReallyReallyReallyReallyReallyReallyReallyReallyLongDetails
    Keys: 
      Id PK
  EntityType: Location
    Properties: 
      AReallyReallyReallyReallyReallyReallyReallyReallyReallyLongDetailsId (no field, int) Shadow Required PK FK AfterSave:Throw ValueGenerated.OnAdd
      Lat (float) Required
      Long (float) Required
    Keys: 
      AReallyReallyReallyReallyReallyReallyReallyReallyReallyLongDetailsId PK
    Foreign keys: 
      Location {'AReallyReallyReallyReallyReallyReallyReallyReallyReallyLongDetailsId'} -> AReallyReallyReallyReallyReallyReallyReallyReallyReallyLongDetails {'Id'} Unique Ownership ToDependent: Location Cascade
```

**Comment by reinux** [comment]
Been swamped lately... Will get back to you as soon as I have something.

**Comment by Arminkhodaei** [comment]
I encountered this problem and this was the reason in my case:

There's a limit to the length of identifiers depending on the underlying database. In my case (postgres) the identifiers length is limited to 63 bytes by default and the issue occurs as auto generated table names from the owned types can exceed this limits.

**Solution:**
I solved the issue by simply setting a custom and shorter table name on the owned collection.

References:
PostgreSQL: 
https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS

SQLServer:
https://docs.microsoft.com/en-us/sql/relational-databases/databases/database-identifiers?view=sql-server-ver16#classes-of-identifiers
